### PR TITLE
stricter behavior for parse_json, add try_parse_json, remove to_json

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
This PR adjusts the behavior of `PARSE_JSON` to be strict and a bit more intuitive, adds a new `TRY_PARSE_JSON` which will produce `NULL` values instead of errors for bad or non-string inputs, and removes `TO_JSON` which no longer had a use since its behavior has been made intrinsic to all JSON expressions.

Previously `PARSE_JSON` was far far too forgiving, allowing stuff like `PARSE_JSON('hello world')` to pass through into a string root literal even though it isn't valid JSON. Meanwhile, something like `PARSE_JSON('{')` would be an error. Basically, the behavior was `TO_JSON` if the input was not actually JSON, which is far too magical and not particularly useful (apologies this made it in like this, it was a relic of an earlier iteration).

Now, `PARSE_JSON` is strict and will fail with an error if provided non-string input (with the exception of null, which will output a null) or if the string contains invalid JSON. `TRY_PARSE_JSON` has also been added, which will return `NULL` for anything that is not a string containing valid JSON in the cases where `PARSE_JSON` would have resulted in an error.

I made some adjustments to `StructuredDataProcessor` to look for and 'unwrap' any `StructuredData` objects it happens to encounter, which would allow for the JSON expressions which produce `COMPLEX<json>` outputs to return them as `StructuredData`, but have not actually made this change to the expressions at this time. This change should be harmless, but isn't currently necessary so I can revert it if any reviewers wish.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by vogievetsky** [comment]
I have not had the pleasure yet of checking out this branch and playing with it but from the description provided it sounds great. Very excited for `TRY_PARSE_JSON`!

File: processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java [hunk 1]
```diff
@@ -139,38 +149,45 @@ public ToJsonExpr(List<Expr> args)
...
-      return new ToJsonExpr(args);
+      return new ToJsonStringExpr(args);
     }
   }
 
-  public static class ToJsonStringExprMacro implements ExprMacroTable.ExprMacro
+  public static class ParseJsonExprMacro implements ExprMacroTable.ExprMacro
   {
-    public static final String NAME = "to_json_string";
+    public static final String NAME = "parse_json";
```
Comment (cryptoe) [inline processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java:185]: Is there a doc update required for this ?

**Review by cryptoe — APPROVED** [review]
LGTM one minor comment regarding docs. 
+1 non binding

File: processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java (hunk 1 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java:185]: no, docs in #12922 were written as if this PR were already merged, (`to_json_string` still exists, but `to_json` is removed and `try_parse_json` is added, so this is a result of funny diff stuff i think)

**Review by gianm — APPROVED** [review]
LGTM. The new behavior is more internally consistent, and is therefore an improvement.

**Pull request merged**